src/dig_hole_test.cc

```diff
@@ -0,0 +1,236 @@
+
+  void DelKeyValue(const std::string &key) {
+    std::string key_str = EncodeKey(key);
+    data_.erase(key_str);
+  }
+
+  void NewBuilder() {
+    TitanDBOptions db_options(titan_options_);
+    TitanCFOptions cf_options(titan_options_);
+    BlobFileCache cache(db_options, cf_options, {NewLRUCache(128)}, nullptr);
+    {
+      std::unique_ptr<WritableFile> f;
+      ASSERT_OK(env_->NewWritableFile(file_name_, &f, env_options_));
+      writable_file_.reset(
+          new WritableFileWriter(std::move(f), file_name_, env_options_));
+    }
+    builder_.reset(
+        new BlobFileBuilder(db_options, cf_options, writable_file_.get()));
+    data_.clear();
+  }
+
+  void FinishBuilder() {
+    ASSERT_OK(builder_->Finish());
+    ASSERT_OK(builder_->status());
+  }
+
+  void CheckKeyExists() {
+    // open file
+    uint64_t file_size = 0;
+    ASSERT_OK(env_->GetFileSize(file_name_, &file_size));
+    OpenBlobFile(file_number_, 0, titan_options_, env_options_, env_,
+                 &random_rw_file_);
+    blob_file_iterator_.reset(new BlobFileIterator{
+        std::move(random_rw_file_), file_number_, file_size, TitanCFOptions()});
+    // get keys in file
+    std::unordered_set<std::string> keys_in_file;
+    for (blob_file_iterator_->SeekToFirst(); blob_file_iterator_->Valid();
+         blob_file_iterator_->Next()) {
+      ASSERT_OK(blob_file_iterator_->status());
+      std::string key_str = DecodeKey(blob_file_iterator_->key());
+      keys_in_file.insert(key_str);
+    }
+    // data_ should be subset of keys_in_file
+    for (auto &iter : data_) {
+      bool ans = keys_in_file.find(iter.first) != keys_in_file.end();
+      if (!ans) {
+        LOG(INFO) << iter.first;
+        assert(ans);
+      }
+    }
+  }
+  //Write random number and random length records to blob file, delete random number records and dig.
+  //Test size before dig and size after dig, and then check remain key in the file.
+  void Test(uint64_t threshold_discard) {
+    NewBuilder();
+    // add records
+    int32_t n = Random() * kRecordNum / kRandomMax + 1;
+    std::vector<BlobHandle> handles(n);
+    for (int32_t i = 0; i < n; i++) {
+      auto id = std::to_string(i);
+      std::string value =
+          std::string(Random() * kValueMaxLength / kRandomMax, 'v');
+      AddKeyValue(id, value, &handles[i]);
+      assert(handles[i].offset < kBlobMaxFileSize);
+    }
+    FinishBuilder();
+    GetExpectBeforeSize(handles[n - 1]);
+    // del records
+    for (int32_t i = 0; i < n; i++) {
+      auto id = std::to_string(i);
+      if (Random() > threshold_discard) {
+        DelKeyValue(id);
+      }
+    }
+    // check before size
+    uint64_t file_size = 0;
+    ASSERT_OK(env_->GetFileSize(file_name_, &file_size));
+    uint64_t before_size = 0;
+    GetRealSize(&before_size);
+    ASSERT_EQ(before_size, expect_before_size);
+    // dig
+    BlobFileMeta blob_file_meta(file_number_, file_size);
+    Status s = dig_hole_job_->Exec(&blob_file_meta);
```

Comment:
Please check the real_file_size in BlobFileMeta

Reply:
Done.